[C#] time new exercise

mikedamay opened the pull request:
Please review Example.cs, design.md and instructions.md

All other files apart from _hints.md_ are complete subject to proofing and show the direction of travel.

mikedamay commented:
/format
/format-csharp

ErikSchierboom commented on languages/csharp/exercises/concept/time/Time.csproj:
```diff
@@ -0,0 +1,18 @@
+        <TargetFramework>netcoreapp3.1</TargetFramework>
+        <IsWindows Condition="'$([System.Runtime.InteropServices.RuntimeInformation]::IsOSPlatform($([System.Runtime.InteropServices.OSPlatform]::Windows)))' == 'true'">true</IsWindows>
+    </PropertyGroup>
+
+    <PropertyGroup Condition="'$(IsWindows)'=='true'">
+        <DefineConstants>Windows</DefineConstants>
+    </PropertyGroup>
```
I know I'm not supposed to be reading this line, but what does this do?

ErikSchierboom commented on languages/csharp/exercises/concept/time/Time.csproj:
```diff
@@ -0,0 +1,18 @@
+
+    <PropertyGroup>
+        <TargetFramework>netcoreapp3.1</TargetFramework>
+        <IsWindows Condition="'$([System.Runtime.InteropServices.RuntimeInformation]::IsOSPlatform($([System.Runtime.InteropServices.OSPlatform]::Windows)))' == 'true'">true</IsWindows>
```
I know I'm not supposed to be reading this line, but what does this do?

ErikSchierboom commented on languages/csharp/exercises/concept/time/.meta/Example.cs:
```diff
@@ -0,0 +1,128 @@
+        return new CultureInfo(cultureId);
+    }
+
+#if Windows
```
I have a thing against `#if`-defs. Can you perhaps make it such that it doesn't use an if-def? I also don't expect us to take a dependency on defines.

ErikSchierboom replied:
Oh I forgot that you can actually check for the runtime platform in .NET Core:

```csharp
RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? ... : 
```

mikedamay replied:
I'll do it at runtime.

mikedamay replied:
Done

ErikSchierboom commented on languages/csharp/exercises/concept/time/.meta/Example.cs:
```diff
@@ -0,0 +1,128 @@
+#else    
+    private static string GetTimeZoneId(Location location)
+    {
+        string timeZoneId = string.Empty;
```
I think a direct return here would be slightly more concise and idiomatic (I could be wrong of course):

```suggestion
        switch (location)
        {
            case Location.NewYork:
                return "America/New_York";
            case Location.London:
                return "Europe/London";
            case Location.Paris:
                return "Europe/Paris";
            default:
                return string.Empty;
        }
```

ErikSchierboom commented on languages/csharp/exercises/concept/time/.meta/Example.cs:
```diff
@@ -0,0 +1,128 @@
+        }
+        catch (Exception)
+        {
+            return new DateTime();
```
The tests use `DateTime.MinValue`, which you could also use here:

```suggestion
            return DateTime.MinValue;
```

ErikSchierboom commented on languages/csharp/exercises/concept/time/.meta/Example.cs:
```diff
@@ -0,0 +1,128 @@
+    public static bool HasDaylightSavingChanged(DateTime dt, Location location)
+    {
+        DateTime dtPrevious = dt.AddDays(-7);
+        string tzid = string.Empty;
```
I don't think this is used.

```suggestion
```

mikedamay commented:
Closed for proofing - will reopen shortly

ErikSchierboom commented on languages/csharp/exercises/concept/time/.meta/Example.cs:
```diff
@@ -0,0 +1,147 @@
+        return new CultureInfo(cultureId);
+    }
+
+    // ****** we are not exercising cross-platform skills as part of the exercise.
```
We could consider having the time zone ID retrieval methods already stubbed out, because as you said, we are not teaching cross-platform skills. What do you think?

mikedamay replied:
> we are not teaching cross-platform skills

I agree that we don't want the student to start messing about with different platforms but I would suggest it is important that they are aware of this unhappiness relating to time zones and that they get hands-on with the ids which are really one the trickiest things in this area.  I want them to think about ids, that they come from outside of .NET and in the case of Windows are not a very nice format.

Assuming I have understood your position I would still push for them to be doing their own location-to-timezoneId mapping.

ErikSchierboom replied:
> Assuming I have understood your position I would still push for them to be doing their own location-to-timezoneId mapping.

Sure. I'm wondering if we should mention in the `introduction.md` file that there is a `RuntimeInformation` class, just to get them looking in the right direction.

mikedamay replied:
> Sure. I'm wondering if we should mention in the introduction.md file that there is a RuntimeInformation class, just to get them looking in the right direction.

I'm a bit confused now.  My original idea (which the instructions and template code reflect) was that the student would not h be concerned with any cross-platform **technology** and we would provide appropriate time zone ids for their platform (in the instructions).  They would just be aware that there was a potential cross-platform issue.

I suppose there should be a note for students using the in-browser experience.  I'm assuming it is Linux.

I don't mind mentioning cross-platform technology or including it in the exercise.  I just want to put the students in a position where they have to take notice of these ids.  I think if we give them the ids in the code then they will forget about them.

I think we have a number of options including mentioning the `RuntimeInformation` class in the introduction but I would like to unconfused about your position and make sure you understand mine, first.

ErikSchierboom replied:
> I'm a bit confused now. My original idea (which the instructions and template code reflect) was that the student would not h be concerned with any cross-platform technology and we would provide appropriate time zone ids for their platform (in the instructions). They would just be aware that there was a potential cross-platform issue.

Well, it isn't just a potential cross-platform issue, as for CLI users that will be using Windows, their tests will fail when run in the test runner. So we have to do something to get them to write the correct code.

Makes sense?

mikedamay replied:
I hadn't thought of that.  My understanding is now that even if it works on Windows on a local machine it will break when uploaded to the test runner unless some cross-platform capability is built in.

An alternative would be to put the cross-platform logic in the tests but I think that would be worse than in _Time.cs_.

Ok. I can't think of anything better than stubbing out the code and making a fuss about time zone ids in the _instructions.md_  as well as _after.md_.  Unsatisfactory, but that is life.  I will do that and mention the `RuntimeInformation` class in the introduction.

ErikSchierboom replied:
> Ok. I can't think of anything better than stubbing out the code and making a fuss about time zone ids in the instructions.md as well as after.md. Unsatisfactory, but that is life. I will do that and mention the RuntimeInformation class in the introduction.

I'd be totally okay with just mentioning `RuntimeInformation` in the `introduction.md` and then having the students figuring it out themselves (with some relatively spoiler-y hints).

mikedamay replied:
So you're saying let the students do the cross-platform stuff.  No code mapping location to timezone id and no cross-platform code is given away.

That is fine with me.  They get to grapple with the time zone ids.

Do you want to add cross-platform as a concept?

ErikSchierboom replied:
Yes, precisely. If we give them a very nice hint and mention `RuntimeInformation` in the `introduction.md`, I think most of the will be able to figure this out.

mikedamay replied:
Implemented the above.  See comment below.

mikedamay commented:
Augmented _introduction.md for cross-platform, added as a concept to _design.md_, changed wording of task 2 in _instructions.md_ and removed comment from _Example.cs_ that student was not being exercised about cross-platform.
